File: backend/backup/feature-flag-service/go/internal/model/feature_flag.go
package model

import (
	"time"
)

// FeatureFlag represents a feature flag
type FeatureFlag struct {
	Key         string            `json:"key"`
	Namespace   string            `json:"namespace"`
	Value       *FlagValue        `json:"value"`
	Description string            `json:"description,omitempty"`
	Tags        map[string]string `json:"tags,omitempty"`
	Temporary   bool              `json:"temporary,omitempty"`
	ExpiresAt   *time.Time        `json:"expiresAt,omitempty"`
	Version     string            `json:"version,omitempty"`
	CreatedAt   time.Time         `json:"createdAt"`
	UpdatedAt   time.Time         `json:"updatedAt"`
	CreatedBy   string            `json:"createdBy,omitempty"`
	UpdatedBy   string            `json:"updatedBy,omitempty"`
	Rules       []*FlagRule       `json:"rules,omitempty"`
}

// NewFeatureFlag creates a new feature flag
func NewFeatureFlag(key, namespace string, value *FlagValue) *FeatureFlag {
	now := time.Now()
	return &FeatureFlag{
		Key:       key,
		Namespace: namespace,
		Value:     value,
		CreatedAt: now,
		UpdatedAt: now,
		Tags:      make(map[string]string),
	}
}

// IsExpired checks if the flag is expired
func (f *FeatureFlag) IsExpired() bool {
	return f.Temporary && f.ExpiresAt != nil && time.Now().After(*f.ExpiresAt)
}

// SetExpiration sets the expiration time for a temporary flag
func (f *FeatureFlag) SetExpiration(duration time.Duration) {
	f.Temporary = true
	expiresAt := time.Now().Add(duration)
	f.ExpiresAt = &expiresAt
}

// AddTag adds a tag to the flag
func (f *FeatureFlag) AddTag(key, value string) {
	if f.Tags == nil {
		f.Tags = make(map[string]string)
	}
	f.Tags[key] = value
}

// RemoveTag removes a tag from the flag
func (f *FeatureFlag) RemoveTag(key string) {
	if f.Tags != nil {
		delete(f.Tags, key)
	}
}

// AddRule adds a rule to the flag
func (f *FeatureFlag) AddRule(rule *FlagRule) {
	if f.Rules == nil {
		f.Rules = make([]*FlagRule, 0)
	}
	f.Rules = append(f.Rules, rule)
}

// RemoveRule removes a rule from the flag
func (f *FeatureFlag) RemoveRule(ruleID string) {
	if f.Rules == nil {
		return
	}
	
	for i, rule := range f.Rules {
		if rule.ID == ruleID {
			f.Rules = append(f.Rules[:i], f.Rules[i+1:]...)
			return
		}
	}
}

File: backend/backup/feature-flag-service/python/feature_flag_service/model/feature_flag.py
"""
Feature flag module for the Feature Flag Service.
"""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from feature_flag_service.model.flag_rule import FlagRule
from feature_flag_service.model.flag_value import FlagValue


class FeatureFlag(BaseModel):
    """
    Model for a feature flag.
    """
    key: str
    namespace: str
    value: FlagValue
    description: Optional[str] = None
    tags: Dict[str, str] = Field(default_factory=dict)
    temporary: bool = False
    expires_at: Optional[datetime] = None
    version: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    rules: List[FlagRule] = Field(default_factory=list)

    def is_expired(self) -> bool:
        """
        Check if the flag is expired.
        """
        return self.temporary and self.expires_at is not None and datetime.now() > self.expires_at

    def set_expiration(self, duration_seconds: int) -> None:
        """
        Set the expiration time for a temporary flag.
        """
        self.temporary = True
        self.expires_at = datetime.now().timestamp() + duration_seconds

    def add_tag(self, key: str, value: str) -> None:
        """
        Add a tag to the flag.
        """
        self.tags[key] = value

    def remove_tag(self, key: str) -> None:
        """
        Remove a tag from the flag.
        """
        if key in self.tags:
            del self.tags[key]

    def add_rule(self, rule: FlagRule) -> None:
        """
        Add a rule to the flag.
        """
        self.rules.append(rule)

    def remove_rule(self, rule_id: str) -> None:
        """
        Remove a rule from the flag.
        """
        self.rules = [rule for rule in self.rules if rule.id != rule_id]

